exercise/cache/points.py

```diff
@@ -274,58 +348,69 @@ def get_max_deviations(
                 )
 
             # Augment exercise reveal rules.
-            if not self.is_staff:
-                for exercise in (
-                    BaseExercise.objects
-                    .filter(course_module__course_instance=instance)
-                    .prefetch_related('submission_feedback_reveal_rule')
-                    .only('id', 'submission_feedback_reveal_rule')
-                ):
-                    try:
-                        tree = self._by_idx(modules, exercise_index[exercise.id])
-                    except KeyError:
-                        self.dirty = True
-                        continue
-                    entry = tree[-1]
-                    rule = exercise.active_submission_feedback_reveal_rule
-                    state = ExerciseRevealState(entry)
-                    is_revealed = rule.is_revealed(state)
-                    reveal_time = rule.get_reveal_time(state)
-
+            for exercise in (
+                BaseExercise.objects
+                .filter(course_module__course_instance=instance)
+                .prefetch_related('submission_feedback_reveal_rule')
+                .only('id', 'submission_feedback_reveal_rule')
+            ):
+                try:
+                    tree = self._by_idx(modules, exercise_index[exercise.id])
+                except KeyError:
+                    self.dirty = True
+                    continue
+                entry = tree[-1]
+                rule = exercise.active_submission_feedback_reveal_rule
+
+                # Evaluate the reveal rule using the real (staff) values.
+                entry_copy = dict(entry)
+                del entry_copy['submissions'] # Not needed by the reveal rule and we don't want to unprefix them.
+                self._unprefix(entry_copy, True)
+                state = ExerciseRevealState(entry_copy)
+                is_revealed = rule.is_revealed(state)
+                reveal_time = rule.get_reveal_time(state)
+
+                entry.update({
+                    '_student_feedback_revealed': is_revealed,
+                    '_student_feedback_reveal_time': reveal_time,
+                })
```

Comment:
Why is the copied entry needed for the reveal rules? I'm also confused of the need to unprefix it here.

Reply:
The reason is that the `ExerciseRevealState` initializer expects a fully formed `CachedPoints` entry (i.e. one that has been received by constructing a `CachedPoints` and has been unprefixed by the `CachedPoints` initializer). However, this section of code is a bit special, because we're initializing an `ExerciseRevealState` **before** the cache generation has completed. Therefore we copy the `entry` in its current state and unprefix it, so that `ExerciseRevealState` can handle it.